text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+        { .. }
+
+    // Chain a future onto this one
+    fn then<F>(self, f: impl FnOnce(Self::Item) -> F) -> impl Future<Item = F::Item>
+        where F: Future
+        { .. }
+
+    // Chain a closure for side effects
+    fn inspect(self, f: impl FnOnce(&Self::Item)) -> impl Future<Item = Self::Item>
+        { .. }
+
+    // Translate unwinding within this future into a `Result`
+    fn catch_unwind(self) -> impl Future<Item = Result<Self::Item, Box<Any + Send>>>
+        where Self: UnwindSafe
+        { .. }
+}
+```
+
+These adapters are all straightforward and based on substantial prior art.
+
+More interesting adapters like `shared`, `select` and `join` will be left to the
+ecosystem to iterate on before being RFC'ed for `std`.
+
+### `Result`-specific adapters
+
+Futures are often enough used with `Result` values that we provide a distinct
+subtrait for that case, equipped with some additional adapters:
+
+```rust
+trait FutureRes<T, E>: Future<Item = Result<T, E>> {
+    // Transform the successful result of the future
+    fn map_ok<U>(self, f: impl FnOnce(T) -> U) -> impl FutureRes<U, E>
+        { .. }
+
+    // Transform the error result of the future
+    fn map_err<F>(self, f: impl FnOnce(E) -> F) -> impl FutureRes<T, F>
+        { .. }
+
+    // Chain a future onto this one on success
+    fn and_then<F, U>(self, f: impl FnOnce(T) -> F) -> impl FutureRes<U, E>
+        where F: FutureRes<U, E>
+        { .. }
+
+    // Chain a future onto this one on failure
+    fn or_else<F, G>(self, f: impl FnOnce(E) -> F) -> impl FutureRes<T, G>
+        where F: FutureRes<T, G>
+        { .. }
+
+    // Pass the error type through an arbitrary conversion
+    fn err_into<F>(self) -> impl FutureRes<T, F>
+        where E: Into<F>
+        { .. }
+
+    // Handle the error provided by this future
+    fn recover<F>(self, f: impl FnOnce(E) -> F) -> impl Future<Item = T>
+        where F: Future<Item = T>
+        { .. }
+}
+
+// Automatically applied to all `Result`-returning futures
+impl<T, E, F> FutureRes<T, E> for F where F: Future<Item = Result<T, E>> {}
+```
+
+## Prelude
+
+The `Future` and `FutureRes` traits are added to the prelude.
+
+## Stabilization plan
+
+The holy grail would be to stabilize async/await for Rust 2018 (roughly by mid-September).
+
+As of this writing, the futures crate is just about to release its 0.2 version;
+some details are available [here](http://aturon.github.io/2018/02/27/futures-0-2-RC/).
+
+The APIs proposed here roughly correspond to the `futures-core` part of the
+crate, plus some adapters that are currently within `futures-util`. However,
+there are two substantial changes in this RFC:
+
+- The use of [pinned types] to enable borrowing within futures.
+  - It's currently not possible to use this API due to rustc limitations; these
+    are expected to be addressed very soon.
+
+- Removing the associated `Error` type (and adjusting combinators accordingly).
```

Comment:
All of our Futures use cases at OneSignal have some sort of an `Error` variant. This is because all of our uses of futures are around asynchronous I/O which is of course fallible. The other use case for futures is to offload some CPU task to a background thread, but there's still possibility for error in those cases since _threads can panic_.

I don't understand the motivation to remove the `Error` associated type. How is this "long desired"? Who desires it? This is making the 99% use case take a back seat to an unclear use case.